codegen/java/src/main/kotlin/io/spine/protodata/codegen/java/CodeLayout.kt

```diff
@@ -76,43 +78,70 @@ public fun EnumType.javaClassName(declaredIn: File): ClassName =
     name.javaClassName(declaredIn)
 
 /**
- * Obtains the full name of the Java class, generated from the Protobuf type with this name.
+ * Obtains a Java class name for the given Protobuf declaration.
  *
- * @return name of the Java class.
+ * @param declaredIn
+ *        the Protobuf file where the declaration is declared.
+ * @param block
+ *        of code which adds the name elements to the class name.
  */
-internal fun TypeName.javaClassName(declaredIn: File): ClassName {
+private fun composeJavaClassName(
+    declaredIn: File,
+    block: MutableList<String>.() -> Unit
+): ClassName {
     val packageName = declaredIn.javaPackage()
     val javaMultipleFiles = declaredIn.javaMultipleFiles()
     val nameElements = mutableListOf<String>()
     if (!javaMultipleFiles) {
         nameElements.add(declaredIn.javaOuterClassName())
     }
-    nameElements.addAll(nestingTypeNameList)
-    nameElements.add(simpleName)
+    block(nameElements)
     return ClassName(packageName, nameElements)
 }
 
+/**
+ * Obtains a fully-qualified Java class, generated for the Protobuf type with this name.
+ */
+internal fun TypeName.javaClassName(declaredIn: File): ClassName =
+    composeJavaClassName(declaredIn) {
+        addAll(nestingTypeNameList)
+        add(simpleName)
+    }
+
+/**
+ * Obtains a fully-qualified Java class, generated for the gRPC service with this name.
+ */
+internal fun ServiceName.javaClassName(declaredIn: File): ClassName =
+    composeJavaClassName(declaredIn) {
+        add(simpleName + "Grpc")
```

Comment:
Continuing on my previous comment. Again, this may not be a gRPC service. Let's discuss this edge case (like the one our guys have in our products) vocally.